server/plugin_release.go

```diff
@@ -4,396 +4,699 @@
+	}
+
+	// Upload github plugin tar signature to github
+	githubPluginSignatureFilePath := githubPluginFilePath + ".sig"
+	if err := uploadFilesToGithub(ctx, client, owner, repositoryName, tag, []string{githubPluginSignatureFilePath}); err != nil {
+		return errors.Wrapf(err, "failed to upload files to github")
+	}
+
+	// Duplicate github plugin tar and its signature that follows s3 release bucket naming convention
+	s3PluginFilepath := filepath.Join(tmpFolder, fmt.Sprintf("%v-%v.tar.gz", repositoryName, tag))
+	if err := os.Symlink(githubPluginFilePath, s3PluginFilepath); err != nil {
+		return errors.Wrapf(err, "failed to duplicate plugin file")
+	}
 
+	s3PluginSignatureFilepath := s3PluginFilepath + ".sig"
+	if err := os.Symlink(githubPluginSignatureFilePath, s3PluginSignatureFilepath); err != nil {
+		return errors.Wrapf(err, "failed to duplicate signature file")
+	}
+
+	s3Bucket := []string{s3PluginFilepath, s3PluginSignatureFilepath}
+	for _, p := range platformPluginFilePaths {
+		s3Bucket = append(s3Bucket, p)
+		s3Bucket = append(s3Bucket, fmt.Sprintf("%s.sig", p))
+	}
+
+	// Upload plugins and signatures to s3 release bucket
+	if err := uploadToS3(ctx, Cfg, s3Bucket); err != nil {
+		return errors.Wrapf(err, "failed to upload to s3")
 	}
 
 	return nil
 }
 
-func createTag(tag, repo string) error {
-	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: Cfg.GithubAccessToken})
-	tc := oauth2.NewClient(ctx, ts)
-	client = github.NewClient(tc)
+func checkRepo(ctx context.Context, client *github.Client, owner, repo string) error {
+	result, _, err := client.Search.Repositories(ctx, fmt.Sprintf("repo:%s/%s", owner, repo), nil)
+	if err != nil {
+		return errors.Wrapf(err, "failed to fetch github repo %s", repo)
+	}
+
+	if result.GetTotal() == 0 {
+		return errors.Errorf("looks like this repository is not part of the org or does not exist. Repo: %s", repo)
+	}
 
-	refs, _, err := client.Git.ListRefs(ctx, Cfg.GithubOrg, repo, nil)
+	return nil
+}
+
+func getReleaseByTag(ctx context.Context, client *github.Client, owner, repositoryName, tag string) (*github.RepositoryRelease, error) {
+	release, _, err := client.Repositories.GetReleaseByTag(ctx, owner, repositoryName, tag)
 	if err != nil {
-		return err
+		return nil, errors.Wrapf(err, "failed to get release by tag")
 	}
-	checkTag := fmt.Sprintf("refs/tags/%s", tag)
-	for _, ref := range refs {
-		if ref.GetRef() == checkTag {
-			return fmt.Errorf("Tag %s already exist. Please use another one", tag)
+
+	return release, nil
+}
+
+// createTag creates a new tag at master for repo.
+// Returns ErrTagExists if tag already exists, nil if successful and an error otherwise.
+func createTag(ctx context.Context, client *github.Client, owner, tag, repository string) error {
+	refs, _, err := client.Git.GetRefs(ctx, owner, repository, fmt.Sprintf("tags/%s", tag))
+	if err != nil {
+		var gerr *github.ErrorResponse
+		if !errors.As(err, &gerr) || gerr.Response.StatusCode != http.StatusNotFound {
```

Comment:
Nit: I find this hard to read and understand. Maybe an `else` would make it more clear:
```go
if errors.As(err, &gerr) && gerr.Response.StatusCode == http.StatusNotFound {
	LogInfo("tag %s was not found, creating tag", tag)
} else {
	return errors.Wrapf(err, "failed to get github tag")
}